File: snippets_python/unionfind.py
class UnionFind:
    def __init__(self, size: int):
        self.data = [-1] * size

    def find(self, v: int) -> int:
        if self.data[v] < 0:
            return v

        self.data[v] = self.find(self.data[v])
        return self.data[v]

    def issame(self, u: int, v: int) -> bool:
        return self.find(u) == self.find(v)

    def unite(self, u: int, v: int):
        u = self.find(u)
        v = self.find(v)

        if u == v:
            return

        if self.data[u] > self.data[v]:
            (u, v) = (v, u)

        self.data[u] += self.data[v]
        self.data[v] = u

    def size(self, v: int) -> int:
        return -self.find(self.data[v])

File: src/union_find.rs
pub struct UnionFind {
    data: Vec<i32>,
}

impl UnionFind {
    pub fn new(size: usize) -> Self {
        return Self {
            data: vec![-1; size],
        };
    }

    fn _find(&mut self, v: usize) -> usize {
        assert!(v < self.data.len());
        if self.data[v] < 0 {
            return v;
        }

        self.data[v] = self._find(self.data[v] as usize) as i32;
        return self.data[v] as usize;
    }

    pub fn is_same(&mut self, u: usize, v: usize) -> bool {
        assert!(v < self.data.len() && u < self.data.len());
        self._find(u) == self._find(v)
    }

    pub fn unite(&mut self, mut a: usize, mut b: usize) -> () {
        assert!(a < self.data.len() && b < self.data.len());
        a = self._find(a);
        b = self._find(b);

        if a == b {
            return;
        }
        if self.data[a] > self.data[b] {
            (a, b) = (b, a);
        }

        self.data[a] += self.data[b];
        self.data[b] = a as i32;
    }

    pub fn size(&mut self, mut v: usize) -> i32 {
        assert!(v < self.data.len());
        v = self._find(v);
        -self.data[v]
    }
}
